### PR TITLE
Factor symbol analysis in the bridge from how it is used.

### PR DESCRIPTION
**Pull request by schweitzpgi**

### CODE REVIEW

**Review by jeanPerier — COMMENTED** [review]
While it simplifies the individual pieces of code and add a clean abstraction, I feel like this is adding way too much code and may fire back a lot at code maintenance.
Here is an incomplete counter proposal to this refactoring (I did not even try to compile it, it just shows how I think we could organize the analyzer class to avoid most of the duplication): https://github.com/flang-compiler/f18-llvm-project/commit/4df47aeecf563bec8cb91d8985f791c01b3d7ac8. I think it can be completed in less than 400 lines that would in my opinion be a win.

File: flang/lib/Lower/CallInterface.cpp [hunk 1]
```diff
@@ -236,8 +236,8 @@ void Fortran::lower::CallInterface<T>::mapPassedEntities() {
   if constexpr (std::is_same_v<T, Fortran::lower::CalleeInterface>) {
     assert(inputs.size() == func.front().getArguments().size() &&
            "function previously created with different number of arguments");
-    for (const auto &pair : llvm::zip(inputs, func.front().getArguments()))
-      mapBackInputToPassedEntity(std::get<0>(pair), std::get<1>(pair));
+    for (auto [fst,snd] : llvm::zip(inputs, func.front().getArguments()))
```
[the remark this answers is not in the document]
Comment (schweitzpgi) [reply flang/lib/Lower/CallInterface.cpp:239]: The `const` is actually incorrect. `zip` returns an object here. And on the Mac, every use of `const` in these sorts of idioms is flagged with a warning or three.

**Comment by schweitzpgi** [comment]
> While it simplifies the individual pieces of code and add a clean abstraction, I feel like this is adding way too much code and may fire back a lot at code maintenance.
> Here is an incomplete counter proposal to this refactoring (I did not even try to compile it, it just shows how I think we could organize the analyzer class to avoid most of the duplication): [4df47ae](https://github.com/flang-compiler/f18-llvm-project/commit/4df47aeecf563bec8cb91d8985f791c01b3d7ac8). I think it can be completed in less than 400 lines that would in my opinion be a win.

I should've mentioned that this is just part 1. I haven't finished cleaning everything up or separating out the different use cases. That'll be next. This PR turns the analysis/operation structure "inside out" and has been tested so that there are no known regressions.

File: flang/lib/Lower/Bridge.cpp [hunk 2]
```diff
@@ -2050,94 +1940,431 @@ class FirConverter : public Fortran::lower::AbstractConverter {
...
+    const auto replace = isDummy || isResult;
+    const auto isHostAssoc =
+        Fortran::semantics::IsHostAssociated(sym, sym.owner());
     Fortran::lower::CharacterExprHelper charHelp{*builder, loc};
-    SymbolBoxAnalyzer sba(sym);
-    sba.analyze();
+    Fortran::lower::BoxAnalyzer sba;
+    sba.analyze(sym);
+
+    auto genLBoundsAndExtents = [&](auto &&lex, auto &&hex) {
```
[the remark this answers is not in the document]
Comment (schweitzpgi) [reply flang/lib/Lower/Bridge.cpp:1950]: I'll double-check this. (There was some compiler warnings in this area, and this was a black ops workaround at that time...)

File: flang/lib/Lower/CallInterface.cpp [hunk 3]
```diff
@@ -357,20 +357,18 @@ class Fortran::lower::CallInterfaceImpl {
     // Handle arguments
     const auto &argumentEntities =
         getEntityContainer(interface.side().getCallDescription());
-    for (const auto &pair :
-         llvm::zip(procedure.dummyArguments, argumentEntities)) {
-      const auto &dummyCharacteristic = std::get<0>(pair);
+    for (auto pair : llvm::zip(procedure.dummyArguments, argumentEntities)) {
```
Comment (schweitzpgi) [inline flang/lib/Lower/CallInterface.cpp:360]: For this one, my version of clang++ chokes on the `auto [x,y]` syntax.

**Pull request closed without merging**